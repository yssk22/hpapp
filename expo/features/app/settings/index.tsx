import { logLogin, logLogout } from '@hpapp/system/firebase';
import * as logging from '@hpapp/system/logging';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Platform } from 'react-native';

import SettingsAppConfig, { AppConfig, SettingsAppConfigDefault } from './internals/SettingsAppConfig';
import SettingsCurrentUser, { CurrentUser } from './internals/SettingsCurrentUser';
import SettingsList from './internals/SettingsList';
import Provider, { useSettings } from './internals/SettingsProvider';
import SettingsUPFCConfig, { UPFCConfig } from './internals/SettingsUPFCConfig';
import SettingsUserConfig, { UserConfig, SettingsUserConfigDefault } from './internals/SettingsUserConfig';

export type SettingsProviderProps = {
  appConfig?: AppConfig;
  userConfig?: UserConfig;
  currentUser?: CurrentUser;
  upfcConfig?: UPFCConfig;
};

/**
 * A provider to setup the settings context
 */
export function SettingsProvider({
  appConfig,
  userConfig,
  currentUser,
  upfcConfig,
  children
}: SettingsProviderProps & {
  children: React.ReactNode;
}) {
  const [ready, setReady] = useState(false);
  useEffect(() => {
    (async () => {
      if (Platform.OS === 'web') {
        // we expect web is used only for storybook so we set the user token here from the environment variable and enforce the logged in user
        const accessToken = process.env['HPAPP_USER_TOKEN_FOR_JEST'];
        if (accessToken === undefined) {
          throw new Error('HPAPP_USER_TOKEN_FOR_JEST is not set for web');
        }
        await SettingsCurrentUser.save({
          id: '1234',
          username: 'yssk22',
          accessToken
        });
      } else {
        currentUser && (await SettingsCurrentUser.save(currentUser));
        userConfig && (await SettingsUserConfig.save(userConfig));
        appConfig && (await SettingsAppConfig.save(appConfig));
        upfcConfig && (await SettingsUPFCConfig.save(upfcConfig));
      }
      setReady(true);
    })();
  }, []);
  if (!ready) {
    return null;
  }
  return <Provider settings={SettingsList}>{children}</Provider>;
}

/**
 * Get the application configuration
 */
export function useAppConfig() {
  const [appConfig] = useSettings(SettingsAppConfig);
  return useMemo(() => {
    return {
      ...SettingsAppConfigDefault,
      ...appConfig
    };
  }, [appConfig]);
}

export { SettingsAppConfigDefault };

/**
 * Get the updator for AppConfig
 */
export function useAppConfigUpdator() {
  const [, updator] = useSettings(SettingsAppConfig);
  return updator;
}

/**
 * Get the user configuration
 */
export function useUserConfig() {
  const [userConfig] = useSettings(SettingsUserConfig);
  return userConfig;
}

/**
 * Get the updator for user configuration
 */
export function useUserConfigUpdator() {
  const [, updator] = useSettings(SettingsUserConfig);
  return updator;
}

/**
 * Get the UPFC configuration
 */
export function useUPFCConfig() {
  const [upfcConfig] = useSettings(SettingsUPFCConfig);
  // migrate from legacy helloproject credentials to new one to support mline
  return useMemo(() => {
    if (upfcConfig) {
      return {
        ...upfcConfig,
        username: undefined,
        password: undefined,
        // eslint-disable-next-line deprecation/deprecation
        hpUsername: upfcConfig?.username ?? upfcConfig?.hpUsername,
        // eslint-disable-next-line deprecation/deprecation
        hpPassword: upfcConfig?.password ?? upfcConfig?.hpPassword
      };
    }
    return undefined;
  }, [upfcConfig]);
}

/**
 * Get the updator for UPFC configuration
 */
export function useUPFCConfigUpdator() {
  const [, updator] = useSettings(SettingsUPFCConfig);
  return updator;
}

/**
 * Get the current user
 */
export function useCurrentUser() {
  const [currentUser] = useSettings(SettingsCurrentUser);
  return currentUser;
}

export type User = NonNullable<ReturnType<typeof useCurrentUser>>;

/**
 * Get the updator for current user
 */
export function useCurrentUserUpdator() {
  const [currentUser, updator] = useSettings(SettingsCurrentUser);
  return useCallback(
    (update: CurrentUser | null) => {
      if (currentUser) {
        if (update == null) {
          logging.Info('features.auth.logout', 'logout', {
            userid: currentUser.id
          });
          logLogout();
        }
      } else {
        if (update) {
          logging.Info('features.auth.login', 'login', {
            userid: update.id
          });
          // TODO: we should have a way to know the login method
          logLogin('default', update.id);
        }
      }
      return updator(update);
    },
    [updator]
  );
}

export { SettingsUserConfigDefault };
