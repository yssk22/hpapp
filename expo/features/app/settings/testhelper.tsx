import { act, render } from '@testing-library/react-native';

import { SettingsProvider, SettingsProviderProps } from './';
import SettingsList from './internals/SettingsList';

/**
 * Clear all settings
 */
export async function clearSettings() {
  await Promise.all(
    SettingsList.map((settings) => {
      settings.clear();
    })
  );
}

export async function renderSettingsProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  await clearSettings();
  const content = render(<SettingsProvider {...props}>{children}</SettingsProvider>);
  await act(async () => {});
  return content;
}
