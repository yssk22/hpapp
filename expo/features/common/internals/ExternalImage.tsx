import { Image } from 'expo-image';
import { ComponentProps, useState } from 'react';
import { View } from 'react-native';

const blurhash = 'L:O|w?RkxtWCRkoeayfQ~Ut6Rkoe';

export type ExternalImageProps = Omit<ComponentProps<typeof Image>, 'source'> & {
  spinnerStyle?: React.ComponentProps<typeof View>['style'];
  uri: string;
  /**
   * If true, the image will be cached on the device's file system.
   */
  cachePolicy?: 'none' | 'disk' | 'memory' | 'memory-disk' | null | undefined;
  maxRetries?: number | undefined;
  width: number;
  height: number;
};

/**
 * ExternalImage is an Image component wrapper to display images from external sources.
 * It creates a image cache file on the device's file system to avoid downloading the image multiple times.
 */
export default function ExternalImage({
  uri,
  cachePolicy = 'memory',
  maxRetries = 10,
  spinnerStyle,
  height,
  width,
  style,
  ...rest
}: ExternalImageProps) {
  const [retry, setRetry] = useState(0);
  return (
    <Image
      testID="ExternalImage.Image"
      {...rest}
      cachePolicy={cachePolicy}
      key={`${uri}$$${retry}`}
      source={uri}
      placeholder={{ blurhash }}
      style={[
        {
          width,
          height
        },
        style
      ]}
      transition={1000}
      onError={() => {
        if (retry < maxRetries) {
          setRetry(retry + 1);
        }
      }}
    />
  );
}
