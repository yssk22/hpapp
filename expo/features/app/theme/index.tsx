import { SettingsProvider, SettingsProviderProps } from '@hpapp/features/app/settings';
import { createTheme, calcYIQ, rgbToColor, colorToRGB } from '@hpapp/system/theme/helper';
import { ThemeColorScheme, ThemeColorKey } from '@hpapp/system/theme/types';
import { useTheme } from '@rneui/themed';
import React, { useCallback } from 'react';

import Provider from './internals/ThemeProvider';

export type ThemeProviderProps = SettingsProviderProps;

/**
 * provides the theme context. It has to be under SettingsProvider.
 */
export function ThemeProvider({ children, ...props }: { children: React.ReactNode } & SettingsProviderProps) {
  return (
    <SettingsProvider {...props}>
      <Provider>{children}</Provider>
    </SettingsProvider>
  );
}

const yiqThreshod = 64;

/**
 * get the color and contrast color from the color scheme
 * @param scheme
 * @returns
 */
export function useThemeColor(scheme: ThemeColorScheme) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const def = rgbToColor(color);
  if (def) {
    return [color, def.contrastColor];
  }
  const yiq = calcYIQ(color);
  return [color, yiq > yiqThreshod ? theme.colors.white : theme.colors.black];
}

/**
 * get the color for the skelton
 * @param scheme
 * @param percent
 * @returns
 */
export function useThemeSkeltonColor(scheme: ThemeColorScheme, percent = 0.8) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const [r, g, b] = colorToRGB(color);
  const r1 = Math.min(255, Math.floor(r + (255 - r) * percent));
  const g1 = Math.min(255, Math.floor(g + (255 - g) * percent));
  const b1 = Math.min(255, Math.floor(b + (255 - b) * percent));
  // convert r1, g1, b1 to hex
  return `#${r1.toString(16).padStart(2, '0')}${g1.toString(16).padStart(2, '0')}${b1.toString(16).padStart(2, '0')}`;
}

/**
 * Get the updator function for the theme
 * @returns a function to update the theme
 */
export function useThemeUpdator() {
  const theme = useTheme();
  return useCallback(
    () =>
      function (primaryColor: ThemeColorKey, secondaryColor: ThemeColorKey, bgColor: ThemeColorKey) {
        theme.updateTheme(createTheme(primaryColor, secondaryColor, bgColor));
      },
    [theme.updateTheme]
  );
}
