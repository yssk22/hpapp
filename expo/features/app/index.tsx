import { maybeDev } from '@hpapp/foundation/environment';
import { LocalMediaManagerProvider } from '@hpapp/system/media';

import AppRelayProvider from './internals/AppRelayProvider';
import { default as Root, AppRootProps as RootProps } from './internals/AppRoot';
import AppUpdateBanner from './internals/AppUpdateBanner';
import { ThemeProvider, ThemeProviderProps } from './theme';

export type AppRootProps = RootProps & ThemeProviderProps;

export function AppRoot({ screens, ...props }: AppRootProps) {
  return (
    <ThemeProvider {...props}>
      <AppRelayProvider>
        <LocalMediaManagerProvider name={maybeDev() ? 'hellofanapp-dev' : 'hellofanapp'}>
          <AppUpdateBanner />
          <Root screens={screens} />
        </LocalMediaManagerProvider>
      </AppRelayProvider>
    </ThemeProvider>
  );
}
