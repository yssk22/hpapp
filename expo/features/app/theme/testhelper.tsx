import { SettingsProviderProps } from '@hpapp/features/app/settings';
import { renderSettingsProvider } from '@hpapp/features/app/settings/testhelper';

import { ThemeProvider } from './';

export async function renderThemeProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  return renderSettingsProvider(<ThemeProvider>{children}</ThemeProvider>, props);
}
