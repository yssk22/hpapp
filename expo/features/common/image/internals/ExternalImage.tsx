import { useURICache } from '@hpapp/features/common';
import { ComponentProps, useCallback, useState } from 'react';
import { View, Image, ActivityIndicator, StyleSheet, LayoutChangeEvent } from 'react-native';

export type ExternalImageProps = Omit<ComponentProps<typeof Image>, 'source'> & {
  containerStyle?: React.ComponentProps<typeof View>['style'];
  spinnerStyle?: React.ComponentProps<typeof View>['style'];
  uri: string;
  cache?: boolean;
};

export default function ExternalImage({ containerStyle, width, height, ...rest }: ExternalImageProps) {
  const [dimentions, setDimensions] = useState({ width: 0, height: 0 });
  const onLayout = useCallback((event: LayoutChangeEvent) => {
    const { width, height } = event.nativeEvent.layout;
    setDimensions({ width, height });
  }, []);
  if (width === 0 || height === 0) {
    return <></>;
  }
  return (
    <View style={[containerStyle, { flex: 1, width, height }]} onLayout={onLayout}>
      <ExternalImageInternal {...rest} width={dimentions.width} height={dimentions.height} />
    </View>
  );
}

function ExternalImageInternal({ uri, cache = false, spinnerStyle, ...rest }: ExternalImageProps) {
  const [fallbackFlag, setFallbackFlag] = useState<boolean>(false);
  const localUri = useURICache(uri);
  if (fallbackFlag) {
    return <View>{localUri && <Image {...rest} source={{ uri: localUri }} key={`${localUri}${new Date()}`} />}</View>;
  }
  if (localUri != null) {
    return (
      <Image
        {...rest}
        key={`${localUri!}${new Date()}`}
        source={{ uri: localUri! }}
        onError={() => {
          setFallbackFlag(true);
        }}
      />
    );
  }
  return (
    <View style={[spinnerStyle, styles.spinner]}>
      <ActivityIndicator color="#ff0000" />
    </View>
  );
}

const styles = StyleSheet.create({
  spinner: {
    position: 'absolute',
    alignContent: 'center',
    justifyContent: 'center',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0
  }
});
