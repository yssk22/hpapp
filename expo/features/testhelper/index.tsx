import {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered
} from './internals/jest';
import { ScreenDecorator } from './internals/storybook';

export {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered,
  ScreenDecorator
};
