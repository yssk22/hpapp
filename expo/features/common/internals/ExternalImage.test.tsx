import { renderUserComponent } from '@hpapp/features/app/testhelper';
import {} from '@hpapp/foundation/globals';
import { composeStories } from '@storybook/react';
import { waitFor } from '@testing-library/react-native';

import * as stories from './ExternalImage.stories';

jest.mock('@hpapp/system/uricache', () => {
  const Orig = jest.requireActual('@hpapp/system/uricache');
  return {
    ...Orig,
    getOrCreateCacheURI: jest.fn(async (uri) => {
      return uri;
    })
  };
});

describe('ExternalImage', () => {
  const composed = composeStories(stories);
  Object.keys(composed).forEach((key) => {
    const C = (composed as any)[key];
    test(key, async () => {
      const content = await renderUserComponent(<C />);
      await waitFor(async () => await content.queryByTestId('ExternalImage.Image'));
      expect(content).toMatchSnapshot();
    });
  });
});
