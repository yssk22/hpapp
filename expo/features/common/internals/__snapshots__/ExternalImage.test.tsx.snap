// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalImage Basic 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "#ffffff",
                "flex": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Image
                resizeMethod="resize"
                resizeMode="contain"
                source={
                  {
                    "uri": "http://cdn.helloproject.com/img/rotation/ff25b71d00ea5ff0b44e18c31c0e1e3c4a7230c9.jpg",
                  }
                }
                style={
                  [
                    {
                      "height": 150,
                      "width": 150,
                    },
                    undefined,
                  ]
                }
                testID="ExternalImage.Image"
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitle=""
        backTitleVisible={false}
        backgroundColor="#0075c2"
        color="white"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="/"
        titleColor="white"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
